src/ucm/util/reloc.c

```diff
@@ -302,17 +310,192 @@ static ucs_status_t ucm_reloc_apply_patch(ucm_reloc_patch_t *patch,
+        /* open/close module to get handle */
+        handle = dlopen(ctx.hdr_info.dlpi_name, RTLD_NOW);
+        dlclose(handle);
+        return handle;
+    }
+
+    return NULL;
+}
+
+static char *ucm_reloc_concat_path(const char *dir, const char *file)
+{
+    size_t len;
+    char *path;
+
+    len = strlen(dir);
+    while (len && (dir[len - 1] == '/')) {
+        len--; /* trim closing '/' */
+    }
+
+    while(file[0] == '/') {
+        file++; /* trim beginning '/' */
+    }
+
+    path = malloc(len + strlen(file) + 2);
+    if (path == NULL) {
+        return NULL;
+    }
+
+    memcpy(path, dir, len);
+    path[len] = '/';
+    strcpy(path + len + 1, file);
+
+    return path;
+}
+
+void *ucm_dlopen(const char *filename, int flag)
+{
+    void *handle = NULL;
+    ucm_reloc_patch_t *patch;
+    void *rpath_module;
+    Dl_serinfo stub;
+    Dl_serinfo *serinfo;
+    int res;
+    int i;
+    char *rpath_path;
+    struct stat file_stat;
+
+    ucm_debug("open module: %s, flag: %x", filename, flag);
+
     if (ucm_reloc_orig_dlopen == NULL) {
         ucm_fatal("ucm_reloc_orig_dlopen is NULL");
         return NULL;
     }
 
-    handle = ucm_reloc_orig_dlopen(filename, flag);
+    if (!ucm_global_opts.dlopen_process_rpath) {
+        goto no_rpath;
+    }
+
+    if (filename[0] == '/') { /* absolute path - rpath section is not used */
+        goto no_rpath;
+    }
+
+    rpath_module = ucm_reloc_dlopen_rpath_used(__builtin_return_address(0));
+    if (rpath_module == NULL) { /* rpath section is not found */
+        goto no_rpath;
+    }
+
+    ucm_debug("rpath section found");
+    /* rpath section from module should be used to locate library */
+    res = dlinfo(rpath_module, RTLD_DI_SERINFOSIZE, &stub);
+    if (res) {
+        ucm_debug("dlinfo(RTLD_DI_SERINFOSIZE) failed");
+        goto no_rpath;
+    }
+
+    serinfo = malloc(stub.dls_size);
+    if (serinfo == NULL) {
+        ucm_error("failed to allocate %zu bytes", stub.dls_size);
+        goto no_rpath;
+    }
+
+    *serinfo = stub;
+    res = dlinfo(rpath_module, RTLD_DI_SERINFO, serinfo);
+    if (res) {
+        ucm_debug("dlinfo(RTLD_DI_SERINFO) failed");
+        goto completed_serinfo;
+    }
+
+    for (i = 0; i < serinfo->dls_cnt; i++) {
+        rpath_path = ucm_reloc_concat_path(serinfo->dls_serpath[i].dls_name, filename);
+        if (!rpath_path) {
+            goto completed_serinfo;
+        }
+
+        ucm_debug("check for %s", rpath_path);
+
+        res = stat(rpath_path, &file_stat);
+        if (res) {
+            free(rpath_path);
+            continue;
+        }
+
+        /* ok, file exists, let's try to load it */
+        handle = ucm_reloc_orig_dlopen(rpath_path, flag);
+        free(rpath_path);
+        if (handle != NULL) {
+            /* gotcha */
+            break;
+        }
+    }
+
+completed_serinfo:
+    free(serinfo);
+
+no_rpath:
+    if (handle == NULL) {
+        handle = ucm_reloc_orig_dlopen(filename, flag);
```

Comment:
do this in the beginning of this function, instead of static init

Reply:
added initialization of orig_dlopen into beginning of function